src/binary-reader.cc

```diff
@@ -206,70 +140,207 @@ size_t read_i32_leb128(const uint8_t* p,
+  Result ReadOffset(Offset* offset, const char* desc) WABT_WARN_UNUSED;
+
+  Index NumTotalFuncs();
+  Index NumTotalTables();
+  Index NumTotalMemories();
+  Index NumTotalGlobals();
+
+  Result ReadInitExpr(Index index) WABT_WARN_UNUSED;
+  Result ReadTable(Type* out_elem_type,
+                   Limits* out_elem_limits) WABT_WARN_UNUSED;
+  Result ReadMemory(Limits* out_page_limits) WABT_WARN_UNUSED;
+  Result ReadGlobalHeader(Type* out_type, bool* out_mutable) WABT_WARN_UNUSED;
+  Result ReadFunctionBody(Offset end_offset) WABT_WARN_UNUSED;
+  Result ReadNamesSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadRelocSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadCustomSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadTypeSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadImportSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadFunctionSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadTableSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadMemorySection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadGlobalSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadExportSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadStartSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadElemSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadCodeSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadDataSection(Offset section_size) WABT_WARN_UNUSED;
+  Result ReadSections() WABT_WARN_UNUSED;
+
+  size_t read_end_ = 0; /* Either the section end or data_size. */
+  BinaryReaderDelegate::State state_;
+  BinaryReaderLogging logging_delegate_;
+  BinaryReaderDelegate* delegate_ = nullptr;
+  TypeVector param_types_;
+  std::vector<Index> target_depths_;
+  const ReadBinaryOptions* options_ = nullptr;
+  BinarySection last_known_section_ = BinarySection::Invalid;
+  Index num_signatures_ = 0;
+  Index num_imports_ = 0;
+  Index num_func_imports_ = 0;
+  Index num_table_imports_ = 0;
+  Index num_memory_imports_ = 0;
+  Index num_global_imports_ = 0;
+  Index num_function_signatures_ = 0;
+  Index num_tables_ = 0;
+  Index num_memories_ = 0;
+  Index num_globals_ = 0;
+  Index num_exports_ = 0;
+  Index num_function_bodies_ = 0;
+};
+
+BinaryReader::BinaryReader(const void* data,
+                           size_t size,
+                           BinaryReaderDelegate* delegate,
+                           const ReadBinaryOptions* options)
+    : read_end_(size),
+      state_(static_cast<const uint8_t*>(data), size),
+      logging_delegate_(options->log_stream, delegate),
+      delegate_(options->log_stream ? &logging_delegate_ : delegate),
+      options_(options),
+      last_known_section_(BinarySection::Invalid) {
+  delegate->OnSetState(&state_);
+}
+
+Result WABT_PRINTF_FORMAT(2, 3) BinaryReader::OnError(const char* format, ...) {
+  WABT_SNPRINTF_ALLOCA(buffer, length, format);
```

Comment:
Yeah, in other places I've done
```
PrintError(...)
return Result::Error;
```

I thought this new way was a little nicer since it is a one-liner, but it's less clear. I'll change it.